Control/Monad/Permutations.hs

```diff
@@ -0,0 +1,220 @@
+﻿-- |
+-- Module      :  Control.Monad.Permutations
+-- Copyright   :  © 2017 Mark Karpov
+-- License     :  BSD 3 clause
+--
+-- Maintainer  :  Mark Karpov <markkarpov92@gmail.com>
+-- Stability   :  experimental
+-- Portability :  portable
+--
+-- This module is a generalization of the package @parsec-permutation@
+-- authored by Samuel Hoffstaetter:
+--
+-- http://hackage.haskell.org/package/parsec-permutation
+--
+-- This module also takes inspiration from the algorithm is described
+-- in: /Parsing Permutation Phrases/, by Arthur Baars, Andres Loh and
+-- Doaitse Swierstra. Published as a functional pearl at the Haskell
+-- Workshop 2001.
+--
+-- From these two works we derive a flexible and general method for
+-- parsing permutations over an Applicative structure. Quite useful
+-- in conjunction with \"Free\" constructions of Applicatives, Monads,
+-- etc.
+--
+-- Other permutation parsing libraries tend towards using special \"almost
+-- applicative\" combinators for construction which denies the library user
+-- the ability to lift and unlift permutation parsing into any Applicative
+-- computational context. We redefine these combinators as convenience
+-- operators here alongside the equivalent Applicative instance.
+--
+-- For example, suppose we want to parse a permutation of:
+-- an optional string of @a@'s, the character @b@ and an optional @c@.
+-- Using a standard parsing library combinator @char@, this can be described
+-- using the Applicative instance by:
+--
+-- > test = runPermutation $
+-- >          (,,) <$> toPermutationWithDefault ""  (some (char 'a'))
+-- >               <*> toPermutation (char 'b')
+-- >               <*> toPermutationWithDefault '_' (char 'c')
+--
+-- Equivalently, this can also be describe using the convenience operators
+-- reminiscent of other parsing libraries:
+--
+-- > test = runPermutation $
+-- >          (,,) <$?> ("", some (char 'a'))
+-- >               <||> char 'b'
+-- >               <|?> ('_', char 'c')
+
+module Control.Monad.Permutations
+  (
+  -- ** Permutation type
+    Permutation()
+  -- ** Permutation evaluators
+  , intercalateEffect
+  , runPermutation
+  -- ** Permutation constructors
+  , toPermutation
+  , toPermutationWithDefault
+  -- ** Convenience operators
+  , (<||>)
+  , (<$$>)
+  , (<|?>)
+  , (<$?>)
+  ) where
+
+import Control.Applicative
+
+-- | An Applicative wrapper-type for constructing permutation parsers.
+
+data Permutation m a = P (Maybe a) (m (Permutation m a))
+
+instance Functor m => Functor (Permutation m) where
```

Comment:
Let have indentation as 2 spaces. This will be consistent with the existing code.